pkg/controllers/deprovisioning/multinodeconsolidation.go

```diff
@@ -0,0 +1,153 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deprovisioning
+
+import (
+	"context"
+	"math"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning"
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+)
+
+type MultiNodeConsolidation struct {
+	consolidation
+}
+
+func NewMultiNodeConsolidation(clk clock.Clock, cluster *state.Cluster, kubeClient client.Client, provisioner *provisioning.Provisioner) *MultiNodeConsolidation {
+	return &MultiNodeConsolidation{
+		consolidation{
+			clock:       clk,
+			cluster:     cluster,
+			kubeClient:  kubeClient,
+			provisioner: provisioner,
+		},
+	}
+}
+
+func (m *MultiNodeConsolidation) ComputeCommand(ctx context.Context, candidates ...CandidateNode) (Command, error) {
+	if !m.ShouldAttemptConsolidation() {
+		return Command{action: actionNotPossible}, nil
+	}
+
+	// For now, we will consider up to every node in the cluster, might be configurable in the future.
+	maxParallel := len(candidates)
+	action, err := m.firstNNodeConsolidationOption(ctx, candidates, maxParallel)
+	if err != nil {
+		return Command{}, err
+	}
+	return action, nil
+}
+
+// firstNNodeConsolidationOption looks at the first N nodes to determine if they can all be consolidated at once.  The
+// nodes are sorted by increasing disruption order which correlates to likelihood if being able to consolidate the node
+func (m *MultiNodeConsolidation) firstNNodeConsolidationOption(ctx context.Context, candidates []CandidateNode, max int) (Command, error) {
+	// we always operate on at least two nodes at once, for single nodes standard consolidation will find all solutions
```

Comment:
Is there a reason we can't use multinodeconsolidation to cover the n=1 case?

Reply:
Yeah, multi-node sorts nodes and only looks at a few specific subsets.  Suppose you have consolidatable nodes (c) and non-consolidatable nodes (x).  We sort them by disruption cost, which normally puts the likely consolidatable nodes at the beginning.  However you might have a pod with an instance type selector or an anti-affinity that means a node with few pods can't be consolidated:

```
[x, c, c, x, x, x, x, x] 
```

Multi-node consolidation will fail here as it will only try:

```
[x, c, c, x, x, x, x, x]  => can't consolidate
[x, c, c, x]  => can't consolidate
[x, c] => can't consolidate
```

A single-node consolidation will then work node by node in the same order:
```
[x, c, c, x, x, x, x, x]
 ^ = can't consolidate

[x, c, c, x, x, x, x, x]
    ^ =  consolidate
```


We might be able to improve in the future at maybe sorting these permanently unconsolidatable nodes differently.

Reply:
I wonder if it makes sense to shuffle these a little bit to help find other consolidations. I see what you're saying though.
- Multinodeconsolidation is expensive, so we best effort sort and search in logn time
- Singlenodeconsolidation is less expensive, so we can do a full linear pass

Another thought would be to use a builder approach. e.g.
- sort by likely consolidable
- find a consolidatable node (linearly)
- try to add nodes to the consolidation (linearly)

You could do this in a single linear pass, but you might be able to find bigger matches doing it in n^2. One way would be to bound this by a certain amount of searching time and give up once you hit the timeout.

Reply:
Obviously not blocking -- I think this is good as is, just thinking on it.